File: euler12/euler12.py

def triangle(n):
    return n*(n+1)/2


def count_divisors(n):
    s = 1
    for d in range(1,n/2+1):
        if n % d == 0:
            s += 1
    return s



def tau(n):
    sqroot,t = int(n**0.5),0
    for factor in range(1,sqroot+1):
            if n % factor == 0:
                    t += 2 # both factor and N/factor
    if sqroot*sqroot == n: t = t - 1 # if sqroot is a factor then we counted it twice, so subtract 1
    return t


for n in range(1000000):
    tri  = triangle(n)
    div = tau(tri)
    if div > 500:
        print tri
        print div
        break

File: euler12/euler12.cpp
#include <iostream>
#include <cmath>
using namespace std;


int count_divisors(int n) {
    int count = 0;

    int sqrt_n = sqrt(n);
    for (int i = 1; i <= sqrt_n; i++) {
        if (n % i == 0) {
            count += 2;     // count both i and N/i
        }
    }

    // if n is a perfect square, then we double counted sqrt(n)
    if (sqrt_n*sqrt_n == n) {
        count--;
    }

    return count;
}


int main() {
    int n = 0;

    for (int i = 1;; i++) {
        n += i;
        if (count_divisors(n) >= 500) {
            break;
        }
    }

    cout << n << endl;
}

// 76576500
